hal/inc/ble_hal.h

```diff
@@ -15,190 +15,747 @@
+ *
+ * @param[in]   appearance  Appearance (16-bit), see @ref BLE_APPEARANCES.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_appearance(uint16_t appearance);
 
-// Connection parameters
-typedef struct ble_conn_param {
-    uint16_t version; // API version
-    uint16_t att_mtu_size; // Maximum size of an ATT packet (ATT_MTU)
-} ble_conn_param;
+/**
+ * Get GAP Appearance value.
+ *
+ * @param[out]  appearance  Pointer to appearance (16-bit) to be filled in, see @ref BLE_APPEARANCES.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_get_appearance(uint16_t* appearance);
 
-// Characteristic parameters
-typedef struct ble_char_param {
-    uint16_t version; // API version
-    bool notif_enabled;
-} ble_char_param;
+/**
+ * For Central, it set the connection parameters.
+ * For Peripheral, it set GAP Peripheral Preferred Connection Parameters.
+ *
+ * @param[in]   ppcp    Pointer to a hal_ble_conn_params_t structure with the desired parameters.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_ppcp(const hal_ble_conn_params_t* ppcp, void* reserved);
 
-int ble_init(void* reserved);
+/**
+ * Get GAP Peripheral Preferred Connection Parameters.
+ *
+ * @param[in]   ppcp    Pointer to a hal_ble_conn_params_t structure where the parameters will be stored.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_get_ppcp(hal_ble_conn_params_t* ppcp, void* reserved);
 
-int ble_add_base_uuid(const char* uuid, uint8_t* uuid_type, void* reserved);
+/**
+ * Set a BLE device whitelist. Only one whitelist can be used at a time and the whitelist is shared
+ * between the BLE roles. The whitelist cannot be set if a BLE role is using the whitelist.
+ *
+ * @param[in]   addr_list   Pointer to a whitelist of peer addresses.
+ * @param[in]   len         Length of the whitelist, maximum @ref BLE_GAP_WHITELIST_ADDR_MAX_COUNT.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_add_whitelist(const hal_ble_addr_t* addr_list, uint8_t len, void* reserved);
 
-int ble_init_profile(ble_profile* profile, void* reserved);
+/**
+ * Delete a BLE device whitelist.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_delete_whitelist(void* reserved);
 
-int ble_start_advert(void* reserved);
-void ble_stop_advert(void* reserved);
+/**
+ * Set the TX Power for advertisement.
+ *
+ * @param[in]   value   Radio transmit power in dBm.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gap_set_tx_power(int8_t value);
 
-int ble_get_char_param(uint16_t conn_handle, uint16_t char_handle, ble_char_param* param, void* reserved);
-int ble_set_char_value(uint16_t conn_handle, uint16_t char_handle, const char* data, uint16_t size, unsigned flags, void* reserved);
+/**
+ * Get the TX Power for advertisement.
+ *
+ * @returns     The current TX power.
+ */
+int8_t ble_gap_get_tx_power(void);
```

Comment:
Consider returning `int` here instead, that on error will represent `system_error_t`.

Reply:
Since both the TX power and `system_error_t` are negative value, we cannot distinguish whether the returned value stands for TX power or an error.

Reply:
In this case the signature of the function can be changed as follows:
```cpp
int ble_gap_get_tx_power(int8_t* tx_power, void* reserved);
```